File: bullet.py
"""Bullets fired by both the player and ranged zombies."""
from __future__ import annotations

import math

import pygame

import settings as S
from collision import circle_rect_collide


class Bullet:
    """Simple projectile with trail; handles its own collisions."""

    def __init__(self, pos: pygame.Vector2, angle: float, speed: float,
                 damage: float, owner: str = "player", crit: bool = False,
                 radius: float = 4.0, elem: str | None = None) -> None:
        self.pos = pygame.Vector2(pos)
        self.vel = pygame.Vector2(speed, 0).rotate_rad(angle)
        self.damage = float(damage)
        self.owner = owner                    # 'player' | 'enemy'
        self.crit = crit
        self.radius = radius
        self.lifetime = S.BULLET_LIFETIME
        self.dead = False
        self.elem = elem                      # 'fire' | 'plasma' | 'pierce'
        self.trail_a = pygame.Vector2(pos)
        self.trail_b = pygame.Vector2(pos)
        self.hit_set: set[int] = set()        # for pierce: hit each zombie once

    # ------------------------------------------------------------- update --
    def update(self, dt: float, game) -> None:
        self.lifetime -= dt
        if self.lifetime <= 0:
            self.dead = True
            return

        distance = self.vel.length() * dt
        steps = max(1, int(distance / 10))     # sub-steps avoid tunneling
        step_vec = self.vel * (dt / steps)

        for _ in range(steps):
            self.trail_b.update(self.trail_a)
            self.trail_a.update(self.pos)
            self.pos += step_vec

            if not (0 <= self.pos.x <= S.WORLD_WIDTH and
                    0 <= self.pos.y <= S.WORLD_HEIGHT):
                self.dead = True
                return

            for rect in game.map.get_near(self.pos, self.radius):
                if circle_rect_collide(self.pos.x, self.pos.y, self.radius, rect):
                    self.dead = True
                    game.particles.impact(self.pos, S.color("ui_dim"), 4)
                    return

            if self.owner == "player":
                for z in game.zombies:
                    if id(z) in self.hit_set:
                        continue
                    if self.pos.distance_squared_to(z.pos) <= \
                            (z.radius + self.radius) ** 2:
                        z.take_damage(self.damage, self.crit, game)
                        if self.elem == "fire":
                            game.particles.burn_trail(self.pos,
                                                      math.degrees(
                                                          math.atan2(self.vel.y,
                                                                     self.vel.x)), 4)
                        elif self.elem == "plasma":
                            game.particles.explosion(self.pos, big=False)
                            game.shake_camera(2)
                        if self.elem == "pierce":
                            self.hit_set.add(id(z))
                            continue
                        self.dead = True
                        return
            else:
                p = game.player
                if self.pos.distance_squared_to(p.pos) <= \
                        (p.radius + self.radius) ** 2:
                    p.take_damage(self.damage, game)
                    self.dead = True
                    return

    # --------------------------------------------------------------- draw --
    def draw(self, surface: pygame.Surface, cam) -> None:
        sp = cam.apply(self.pos)
        ta = cam.apply(self.trail_a)
        tb = cam.apply(self.trail_b)
        if self.owner == "player":
            trail_col = (255, 220, 120, 180)
            head_col = (255, 240, 160)
        else:
            trail_col = (255, 100, 90, 180)
            head_col = (255, 180, 150)
        if tb != ta:
            t = max(1, int(self.radius * 1.2))
            pygame.draw.line(surface, trail_col, ta, tb, t)
            inner_t = max(1, int(self.radius * 0.5))
            core_col = (255, 255, 220) if self.owner == "player" else (255, 220, 200)
            pygame.draw.line(surface, core_col, ta, tb, inner_t)
        pygame.draw.circle(surface, head_col, (int(sp.x), int(sp.y)),
                           int(self.radius) + 1)
        pygame.draw.circle(surface, (255, 255, 255),
                           (int(sp.x), int(sp.y)), max(1, int(self.radius) - 2))

File: web/src/game/bullet.ts
// src/game/bullet.ts
// Projectiles fired by player and ranged enemies. Mirrors bullet.py.

import { BULLET_LIFETIME, WORLD_HEIGHT, WORLD_WIDTH } from "./settings";
import { circleRectCollide } from "./collision";
import type { Camera } from "./camera";
import type { Vec } from "./vec";
import type { IGame as Game } from "./types";
import { drawProjectileSprite } from "./pixelArt";

export class Bullet {
  pos: Vec;
  vel: Vec;
  damage: number;
  owner: "player" | "enemy";
  crit: boolean;
  radius: number;
  lifetime: number;
  dead = false;
  trailA: Vec;
  trailB: Vec;

  constructor(
    pos: Vec,
    angle: number,
    speed: number,
    damage: number,
    owner: "player" | "enemy" = "player",
    crit = false,
    radius = 4,
    lifetime = BULLET_LIFETIME,
  ) {
    this.pos = { ...pos };
    this.vel = {
      x: Math.cos(angle) * speed,
      y: Math.sin(angle) * speed,
    };
    this.damage = damage;
    this.owner = owner;
    this.crit = crit;
    this.radius = radius;
    // Weapons can cap their reach per shot (range / speed); enemies and
    // the drone keep the shared default lifetime.
    this.lifetime = lifetime;
    this.trailA = { ...pos };
    this.trailB = { ...pos };
  }

  update(dt: number, game: Game): void {
    this.lifetime -= dt;
    if (this.lifetime <= 0) {
      this.dead = true;
      return;
    }
    const speed = Math.hypot(this.vel.x, this.vel.y);
    const distance = speed * dt;
    const steps = Math.max(1, Math.floor(distance / 10));
    const stepVec = { x: this.vel.x * (dt / steps), y: this.vel.y * (dt / steps) };

    for (let i = 0; i < steps; i++) {
      this.trailB = { ...this.trailA };
      this.trailA = { ...this.pos };
      this.pos.x += stepVec.x;
      this.pos.y += stepVec.y;

      if (
        this.pos.x < 0 ||
        this.pos.x > WORLD_WIDTH ||
        this.pos.y < 0 ||
        this.pos.y > WORLD_HEIGHT
      ) {
        this.dead = true;
        return;
      }

      for (const rect of game.map!.getNear(this.pos, this.radius)) {
        if (circleRectCollide(this.pos.x, this.pos.y, this.radius, rect)) {
          this.dead = true;
          game.particles.impact(this.pos, "#82827E", 4);
          return;
        }
      }

      if (this.owner === "player") {
        for (const z of game.zombies) {
          const dx = this.pos.x - z.pos.x;
          const dy = this.pos.y - z.pos.y;
          const r = (z.radius as number) + this.radius;
          if (dx * dx + dy * dy <= r * r) {
            z.takeDamage(this.damage, this.crit, game);
            this.dead = true;
            game.stats.shots_hit = (game.stats.shots_hit ?? 0) + 1;
            return;
          }
        }
      } else {
        const p = game.player;
        if (!p || p.dead) continue;
        const dx = this.pos.x - p.pos.x;
        const dy = this.pos.y - p.pos.y;
        const r = (p.radius as number) + this.radius;
        if (dx * dx + dy * dy <= r * r) {
          p.takeDamage(this.damage, game);
          this.dead = true;
          return;
        }
      }
    }
  }

  draw(ctx: CanvasRenderingContext2D, cam: Camera): void {
    const sp = cam.apply(this.pos);
    const ta = cam.apply(this.trailA);
    const tb = cam.apply(this.trailB);
    drawProjectileSprite(ctx, sp, ta, tb, this.owner === "enemy");
  }
}
